django/template/loader_tags.py

```diff
@@ -178,12 +178,22 @@ def __init__(self, template, *args, **kwargs):
         super(IncludeNode, self).__init__(*args, **kwargs)
 
     def render(self, context):
+        """
+        Render the specified template and context. Cache the template object
+        in render_context to avoid reparsing and loading when used in a for
+        loop.
+        """
         try:
             template = self.template.resolve(context)
             # Does this quack like a Template?
             if not callable(getattr(template, 'render', None)):
-                # If not, we'll try get_template
-                template = context.template.engine.get_template(template)
+                # If not, we'll try our cache, and get_template
+                template_name = template
+                template = context.render_context.get(template_name)
+                if template is None:
+                    template = context.template.engine.get_template(template_name)
+                context.render_context[template_name] = template
```

Comment:
To aid my understanding, do you have an example of when the same `template_name` would refer to different things? After it's resolved, won't the same name always refer to the same template? Happy to rework with such a cache dict regardless, sharing the same thing across `IncludeNode`s sounds good to me. Can carry on with this in the morning at day 2 :)